webdataset prototype - ReadFilesFromTarIterableDataset

glaringlee opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #50253 [WIP] webdataset prototype - class to preprocess image folder
* #49353 webdataset prototype - GroupByFilenameIterableDataset
* #49140 webdataset prototype - RoutedDecoderIterableDataset
* #49026 webdataset prototype - ReadFilesFromZipIterableDataset
* **#49000 webdataset prototype - ReadFilesFromTarIterableDataset**

Differential Revision: [D25554583](https://our.internmc.facebook.com/intern/diff/D25554583)

Commit:
webdataset prototype - ReadFilesFromTarIterableDataset

[ghstack-poisoned]

Commit:
Update on "[WIP] webdataset prototype - ReadFilesFromTarIterableDataset"




[ghstack-poisoned]

Commit:
Update on "[WIP] webdataset prototype - ReadFilesFromTarIterableDataset"




[ghstack-poisoned]

Commit:
Update on "[WIP] webdataset prototype - ReadFilesFromTarIterableDataset"




[ghstack-poisoned]

Commit:
Update on "[WIP] webdataset prototype - ReadFilesFromTarIterableDataset"




[ghstack-poisoned]

Commit:
Update on "[WIP] webdataset prototype - ReadFilesFromTarIterableDataset"




[ghstack-poisoned]

Commit:
Update on "[WIP] webdataset prototype - ReadFilesFromTarIterableDataset"




[ghstack-poisoned]

Commit:
Update on "[WIP] webdataset prototype - ReadFilesFromTarIterableDataset"




[ghstack-poisoned]

Commit:
Update on "[WIP] webdataset prototype - ReadFilesFromTarIterableDataset"




[ghstack-poisoned]

Commit:
Update on "[WIP] webdataset prototype - ReadFilesFromTarIterableDataset"




[ghstack-poisoned]

Commit:
Update on "[WIP] webdataset prototype - ReadFilesFromTarIterableDataset"




[ghstack-poisoned]

Commit:
Update on "[WIP] webdataset prototype - ReadFilesFromTarIterableDataset"




[ghstack-poisoned]

Commit:
Update on "[WIP] webdataset prototype - ReadFilesFromTarIterableDataset"




[ghstack-poisoned]

Commit:
Update on "[WIP] webdataset prototype - ReadFilesFromTarIterableDataset"




[ghstack-poisoned]

Commit:
Update on "[WIP] webdataset prototype - ReadFilesFromTarIterableDataset"




[ghstack-poisoned]

glaringlee commented on test/test_dataset.py:
```diff
@@ -44,5 +55,23 @@ def test_loadfilesfromdisk_iterable_dataset(self):
+            self.assertTrue(file_pathname in self.temp_files)
+            self.assertTrue(rec[1].read() == open(file_pathname, 'rb').read())
+
+        os.remove(temp_tarfile_pathname)
```
I will remove this line, tear down will remove the entire temp folder including all the files in it.

glaringlee commented on torch/utils/data/datasets/common.py:
```diff
@@ -49,3 +51,62 @@ def get_file_binaries_from_pathnames(pathnames : Iterable):
+                yield (inner_pathname, extract_fobj)
+            return
+    except tarfile.TarError:
+        # Note: We have no way to verify whether a non-seekable stream (eg. PIPE stream) is tar without
```
We do raise the exception if the tar file is corrupted, this is handled by checking the result of `extractfile()`, if it is None, we raise. 
This try block here is not to test whether the tar stream is corrupted, but to test whether the passed in stream is a tar stream, if the passed in stream is not a tar stream, but other format, then we just return the original stream if possible, I am not sure whether we should support this though, but let's say the dataset passed into ReadFilesFromTarIterableDataset is mixed with regular file stream and tar file stream, I think the behavior of ReadFilesFromTarIterableDataset is to un-tar all the tar streams and keep all the non-tar stream the same as it. 
What do you think? Still drop this logic? Please advise.

Commit:
Update on "[WIP] webdataset prototype - ReadFilesFromTarIterableDataset"




[ghstack-poisoned]

Commit:
Update on "[WIP] webdataset prototype - ReadFilesFromTarIterableDataset"




[ghstack-poisoned]

Commit:
Update on "[WIP] webdataset prototype - ReadFilesFromTarIterableDataset"




[ghstack-poisoned]

Commit:
Update on "[WIP] webdataset prototype - ReadFilesFromTarIterableDataset"




[ghstack-poisoned]

Commit:
Update on "[WIP] webdataset prototype - ReadFilesFromTarIterableDataset"




[ghstack-poisoned]

Commit:
Update on "[WIP] webdataset prototype - ReadFilesFromTarIterableDataset"




[ghstack-poisoned]

Commit:
Update on "[WIP] webdataset prototype - ReadFilesFromTarIterableDataset"




[ghstack-poisoned]

glaringlee commented:
Fixed all things that in the comments or discussed previously. Now handles tarfile main stream closed issue correctly.

Commit:
Update on "webdataset prototype - ReadFilesFromTarIterableDataset"




[ghstack-poisoned]

Commit:
Update on "webdataset prototype - ReadFilesFromTarIterableDataset"


Differential Revision: [D25554583](https://our.internmc.facebook.com/intern/diff/D25554583)

[ghstack-poisoned]

Commit:
Update on "webdataset prototype - ReadFilesFromTarIterableDataset"


Differential Revision: [D25554583](https://our.internmc.facebook.com/intern/diff/D25554583)

[ghstack-poisoned]

Commit:
Update on "webdataset prototype - ReadFilesFromTarIterableDataset"


Differential Revision: [D25554583](https://our.internmc.facebook.com/intern/diff/D25554583)

[ghstack-poisoned]

Commit:
Update on "webdataset prototype - ReadFilesFromTarIterableDataset"


Differential Revision: [D25554583](https://our.internmc.facebook.com/intern/diff/D25554583)

[ghstack-poisoned]

glaringlee commented:
updated

VitalyFedyunin commented on torch/utils/data/datasets/readfilesfromtardataset.py:
```diff
@@ -0,0 +1,58 @@
+            warnings.warn("dataset must be Iterable type but got {}".format(type(self.dataset)))
+            raise TypeError
+        for data in self.dataset:
+            ret = validate_pathname_binary_tuple(data)
```
It is better to raise TypeError with proper message out from `validate_pathname_binary_tuple`

VitalyFedyunin commented on torch/utils/data/datasets/readfilesfromtardataset.py:
```diff
@@ -0,0 +1,58 @@
+            ret = validate_pathname_binary_tuple(data)
+            if ret:
+                warnings.warn("got invalid pathname and binary record ({}), abort!".format(ret))
+                raise TypeError
```
You can unpack data into separate vars to make it more readable
```python
file_name, data_stream = data
```

VitalyFedyunin commented on torch/utils/data/datasets/readfilesfromtardataset.py:
```diff
@@ -0,0 +1,58 @@
+                tar = tarfile.open(fileobj=data[1], mode="r:*")
+                for tarinfo in tar:
+                    if not tarinfo.isfile():
+                        continue
```
Curious if it is not the file, what it is? Maybe reason to error?

glaringlee replied:
The original purpose is to bypass dir

VitalyFedyunin commented on torch/utils/data/datasets/readfilesfromtardataset.py:
```diff
@@ -0,0 +1,58 @@
+class ReadFilesFromTarIterableDataset(IterableDataset):
+    r""" :class:`ReadFilesFromTarIterableDataset`.
+
+    IterableDataset to extract tar binary streams from input iterables
```
Please add expected input dataset fields.

VitalyFedyunin commented on torch/utils/data/datasets/readfilesfromtardataset.py:
```diff
@@ -0,0 +1,58 @@
+            dataset : Iterable,
+            length : int = -1):
+        super().__init__()
+        self.dataset : Iterable = dataset
```
Please make Iterable typed here: Iterable[tuple[T, T]] and you can drop entire `validate_pathname_binary_tuple`

glaringlee replied:
will switch typed Iterable but keep validate_pathname_binary_tuple, since type annotation won't really block the execution.

glaringlee commented on test/test_dataset.py:
```diff
@@ -45,5 +56,24 @@ def test_loadfilesfromdisk_iterable_dataset(self):
+            self.assertEqual(rec[1].read(), open(temp_file, 'rb').read())
+        self.assertEqual(count, len(self.temp_files))
+        dataset3.reset()
+
```
I'll add an explicit test here. 
The later group iterable dataset tested this behavior as well though.

Commit:
Update on "webdataset prototype - ReadFilesFromTarIterableDataset"


Differential Revision: [D25554583](https://our.internmc.facebook.com/intern/diff/D25554583)

[ghstack-poisoned]

glaringlee commented:
Updated.
Have to use `typing.cast` to avoid mypy complaining, since tarfile accept `fileobj` as IO[bytes], but our passed in type is `BufferedIOBased` based stream obj. They are compatible, but mypy complains.

Commit:
Update on "webdataset prototype - ReadFilesFromTarIterableDataset"


Differential Revision: [D25554583](https://our.internmc.facebook.com/intern/diff/D25554583)

[ghstack-poisoned]

Commit:
Update on "webdataset prototype - ReadFilesFromTarIterableDataset"


Differential Revision: [D25554583](https://our.internmc.facebook.com/intern/diff/D25554583)

[ghstack-poisoned]

Commit:
Update on "webdataset prototype - ReadFilesFromTarIterableDataset"


Differential Revision: [D25554583](https://our.internmc.facebook.com/intern/diff/D25554583)

[ghstack-poisoned]

VitalyFedyunin reviewed (approved):
Please do not merge with this names, lets wait final go on the naming convention.

glaringlee commented:
Modified to datapipe and merged already
